File: 0006-zigzag-conversion/0006-zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    if(numRows===1 ||s.length <= numRows){
        return s;
    }
    const rows = new Array(numRows).fill('');
    let curRow = 0;
    let movingDown = true;

    for (const char of s){
        rows[curRow]+=char
        
        if(curRow ===0){
            movingDown = true;
        }
        else if (curRow ===numRows-1){
            movingDown = false;
        }
        if(movingDown){
            curRow +=1;
        }
        else{
            curRow -=1;
        }
    }
    return rows.join('')
};

File: 0006-zigzag-conversion/0006-zigzag-conversion.py
class Solution:
    def convert(self,s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s
        
        # Create an array of strings, one for each row
        rows = [''] * numRows
        cur_row = 0
        going_down = False

        for char in s:
            rows[cur_row] += char
            # Change direction when you reach the top or bottom row
            if cur_row == 0 or cur_row == numRows - 1:
                going_down = not going_down
            cur_row += 1 if going_down else -1

        # Combine all rows to form the final output
        return ''.join(rows)
